tests/test_builtin_casters.py

```diff
@@ -251,6 +251,51 @@ def test_integer_casting():
         assert "incompatible function arguments" in str(excinfo.value)
 
 
+def test_int_convert():
+    class DeepThought:
+        def __int__(self):
+            return 42
+
+    class ShallowThought:
+        pass
+
+    class FuzzyThought:
+        def __float__(self):
+            return 41.99999
+
+    convert, noconvert = m.int_passthrough, m.int_passthrough_noconvert
+
+    def require_implicit(v):
+        pytest.raises(TypeError, noconvert, v)
+
+    def cant_convert(v):
+        pytest.raises(TypeError, convert, v)
+
+    assert convert(7) == 7
+    assert noconvert(7) == 7
+    cant_convert(3.14159)
+    assert convert(DeepThought()) == 42
+    require_implicit(DeepThought())
+    cant_convert(ShallowThought())
+    cant_convert(FuzzyThought())
+
+
+def test_numpy_int_convert():
+    np = pytest.importorskip("numpy")
+
+    convert, noconvert = m.int_passthrough, m.int_passthrough_noconvert
+
+    def require_implicit(v):
+        pytest.raises(TypeError, noconvert, v)
+
+    # `np.intc` is an alias that corresponds to a C++ `int`
+    assert convert(np.intc(42)) == 42
+    assert noconvert(np.intc(42)) == 42
+
+    assert convert(np.float32(3.14159)) == 3  # This might be wrong/unwanted?
```

Comment:
This is also kind of weird. It's because we only explicitly test for `float` and not for "float-like" types. I'm not sure if there's a general way to recognize "float-like" types, though. We could just add a special case for NumPy float types?

Not really related to this PR, though, so definitely another PR.

Reply:
I agree, although it looks tricky to implement, therefore I wonder about cost/benefit of working on this.
Pragmatic idea: report the situation in a comment and wait for an opportune moment in the future (e.g. when there is an actual problem that needs to be resolved). My first stab:
```
# The implicit conversion from np.float is undesirable but difficult to detect, without making the pybind11
# int converter dependent on numpy. Currently Python does not have a generalized concept of float-like types.
```

Reply:
We do already have something like this (a special case for NumPy) in the `type_caster` for bool, though.

But yes, this might be something we'll have to live with? At the very least, this should now be considered a "conversion".

Reply:
The Google-global testing of this PR had no issues.
I see two behavior changes here: 1. the one currently implemented, and 2. potentially also raising a TypeError for np.float types (with ifdefs I guess to use the numpy api if available).
Given that this PR is exhaustively tested and doesn't disturb anything anymore, I feel it's best to merge now (for 2.6.2), and try the 2. behavior change in a separate PR. The main rationale for this two-stage approach is to use our time efficiently. Putting off this well-tested change will cost extra time.
For the comment, this is my 2nd suggested version:
```
# The implicit conversion from np.float is undesirable but currently accepted.
```
Just a hint that we thought about it, leaving all options open for the future.

Reply:
Agreed. Never meant to lump this in this PR. I really thought it would just be a straightforward one, but once again, Python had some surprises.

Reply:
Btw, to make things more complicated, `np.float` to `int` isn't allowed, because `np.float` is just a Python `float` (I'm not joking; this is really the case). Anyway, I'll add `32` and commit this suggestion.